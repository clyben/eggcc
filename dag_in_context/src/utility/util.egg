(function ListExpr-length (ListExpr) i64)
(function ListExpr-ith (ListExpr i64) Expr :unextractable)
(function ListExpr-suffix (ListExpr i64) ListExpr :unextractable)
(function Append (ListExpr Expr) ListExpr :unextractable)

(rule ((Switch pred inputs branch)) ((union (ListExpr-suffix branch 0) branch)) :ruleset always-run)

(rule ((= (ListExpr-suffix top n) (Cons hd tl)))
    ((union (ListExpr-ith top n) hd)
     (union (ListExpr-suffix top (+ n 1)) tl)) :ruleset always-run)

(rule ((= (ListExpr-suffix list n) (Nil)))
    ((set (ListExpr-length list) n)) :ruleset always-run)

(rewrite (Append (Cons a b) e)
   (Cons a (Append b e))
   :ruleset always-run)
(rewrite (Append (Nil) e)
   (Cons e (Nil))
   :ruleset always-run)

(function tuple-length (Expr) i64 :unextractable)

(rule ((HasType expr (TupleT tl))
       (= len (TypeList-length tl)))
      ((set (tuple-length expr) len)) :ruleset always-run)

;; Create a Get for every index, and rewrite it to see through Concat
(rule ((Single expr)) ((union (Get (Single expr) 0) expr)) :ruleset always-run)
;; initial get
(rule ((> (tuple-length tuple) 0))
      ((Get tuple 0))
      :ruleset always-run)
;; next get
(rule ((= len (tuple-length tuple))
       (= ith (Get tuple i))
       (< (+ i 1) len)
       )
       ((Get tuple (+ 1 i)))
       :ruleset always-run)

;; descend left
(rule ((Get (Concat expr1 expr2) i)
       (= (tuple-length expr1) len1)
       (< i len1))
      ((union (Get (Concat expr1 expr2) i)
              (Get expr1 i)))
       :ruleset always-run)
;; descend right
(rule ((Get (Concat expr1 expr2) i)
       (= (tuple-length expr1) len1)
       (>= i len1))
      ((union (Get (Concat expr1 expr2) i)
              (Get expr2 (- i len1))))
       :ruleset always-run)


(relation IsLeaf (Expr))
(rule ((Arg ty)) ((IsLeaf (Arg ty))) :ruleset always-run)
(rule ((Const val ty)) ((IsLeaf (Const val ty))) :ruleset always-run)
(rule ((Empty ty)) ((IsLeaf (Empty ty))) :ruleset always-run)

(rewrite (Concat (Concat a b) c)
         (Concat a (Concat b c))
         :ruleset always-run)



;; This Expr size actually do the same thing as cost function.
(function Expr-size (Expr) i64 :unextractable :merge (min old new) )
(function ListExpr-size (ListExpr) i64 :unextractable :merge (min old new))

(rule ((= expr (Function name tyin tyout out)) 
       (= sum (Expr-size out))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (Const n ty))) 
      ((set (Expr-size expr) 1))  :ruleset always-run)

(rule ((= expr (Bop op x y)) 
       (= sum (+ (Expr-size y) (Expr-size x)))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (Uop op x)) 
       (= sum (Expr-size x))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (Get tup i)) 
       (= sum (Expr-size tup))) 
      ((set (Expr-size expr) sum)) :ruleset always-run)

(rule ((= expr (Concat x y)) 
       (= sum (+ (Expr-size y) (Expr-size x)))) 
      ((set (Expr-size expr) sum)) :ruleset always-run)

(rule ((= expr (Single x)) 
       (= sum (Expr-size x))) 
      ((set (Expr-size expr) sum)) :ruleset always-run)

(rule ((= expr (Switch pred branches)) 
       (= sum (+ (ListExpr-size branches) (Expr-size pred)))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (If pred then else)) 
       (= sum (+ (Expr-size else) (+ (Expr-size then) (Expr-size pred))))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (DoWhile in pred-and-output)) 
       (= sum (+ (Expr-size pred-and-output) (Expr-size in)))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (Arg ty))) 
      ((set (Expr-size expr) 1)) :ruleset always-run)

(rule ((= expr (Call func arg)) 
       (= sum (Expr-size arg))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((Empty ty)) ((set (Expr-size (Empty ty)) 0))  :ruleset always-run)

(rule ((= expr (Cons hd tl)) 
       (= sum (+ (ListExpr-size tl) (Expr-size hd)))) 
      ((set (ListExpr-size expr) sum)) :ruleset always-run)

(rule ((Nil)) 
      ((set (ListExpr-size (Nil)) 0))  :ruleset always-run)

(rule ((= expr (Alloc id e state ty)) ;; do state edge's expr should be counted?
        (= sum (Expr-size e))) 
        ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (InContext assumption e)) 
       (= sum (Expr-size e))) 
      ((set (Expr-size expr) sum)) :ruleset always-run)


;;                tuple start len
(function SubTuple (Expr i64 i64) Expr :unextractable)
(rewrite (SubTuple expr x 1)
         (Single (Get expr x)) :ruleset always-run)
(rule ((= f (SubTuple expr a b)) (> b 1))
      ((union f 
         (Concat (Single (Get expr a))
                 (SubTuple expr (+ a 1) (- b 1))))) :ruleset always-run)
