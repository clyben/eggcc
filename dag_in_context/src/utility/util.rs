#[cfg(test)]
use crate::{egglog_test, interpreter::Value};

#[test]
fn test_list_util() -> crate::Result {
    use crate::ast::emptyt;
    let emptyt = emptyt();
    let build = format!(
        "
		(let list (Cons (Const (Int 0) {emptyt} (NoContext))
                  (Cons (Const (Int 1) {emptyt} (NoContext))
                  (Cons (Const (Int 2) {emptyt} (NoContext))
                  (Cons (Const (Int 3) {emptyt} (NoContext))
                  (Cons (Const (Int 4) {emptyt} (NoContext)) (Nil)))))))
		(let expr (Switch (Const (Int 1) {emptyt} (NoContext)) (Empty {emptyt} (NoContext)) list))
	"
    );
    let check = format!(
        "
		(check (= (ListExpr-ith list 1) (Const (Int 1) {emptyt} (NoContext))))
        (check (= (ListExpr-ith list 4) (Const (Int 4) {emptyt} (NoContext))))
        (check (= (ListExpr-length list) 5))
	"
    );
    egglog_test(
        &build,
        &check,
        vec![],
        Value::Tuple(vec![]),
        Value::Tuple(vec![]),
        vec![],
    )
}

#[test]
fn append_test() -> crate::Result {
    use crate::ast::emptyt;
    let emptyt = emptyt();
    let build = format!(
        "
        (let appended
            (Append
                (Cons (Const (Int 0) {emptyt} (NoContext)) (Cons (Const (Int 1) {emptyt} (NoContext)) (Nil)))
                (Const (Int 2) {emptyt} (NoContext))))
    "
    );

    let check = format!("
        (check (
            =
            (Cons (Const (Int 0) {emptyt} (NoContext)) (Cons (Const (Int 1) {emptyt} (NoContext)) (Cons (Const (Int 2) {emptyt} (NoContext)) (Nil))))
            appended
        ))
    ");
    egglog_test(
        &build,
        &check,
        vec![],
        Value::Tuple(vec![]),
        Value::Tuple(vec![]),
        vec![],
    )
}

#[test]
fn test_tuple_ith() -> crate::Result {
    use crate::ast::emptyt;
    let emptyt = emptyt();
    let build = format!(
        "
    (let tup (Concat
                  (Concat (Single (Const (Int 0) {emptyt} (NoContext))) (Single (Const (Int 1) {emptyt} (NoContext))))
                  (Concat (Single (Const (Int 2) {emptyt} (NoContext))) (Single (Const (Int 3) {emptyt} (NoContext))))))
    
    ;; with print
    (let tup2 (Concat 
                (Concat 
                    (Single (Bop (Print) (Const (Int 0) {emptyt} (NoContext)) (Arg (Base (StateT)) (NoContext))))
                    (Concat (Single (Const (Int 1) {emptyt} (NoContext))) 
                                (Single (Const (Int 2) {emptyt} (NoContext)))))
                (Single (Const (Int 3) {emptyt} (NoContext)))))
    "
    );

    let check = format!(
        "
    (check (= (Get tup 0) (Const (Int 0) {emptyt} (NoContext))))
    (check (= (Get tup 1) (Const (Int 1) {emptyt} (NoContext))))
    (check (= (Get tup 2) (Const (Int 2) {emptyt} (NoContext))))
    (check (= (Get tup 3) (Const (Int 3) {emptyt} (NoContext))))
    (check (= 4 (tuple-length tup)))
    (fail (check (Get tup 4)))

    (check (= (Get tup2 0) (Bop (Print) (Const (Int 0) {emptyt} (NoContext)) (Arg (Base (StateT)) (NoContext)))))
    (check (= (Get tup2 1) (Const (Int 1) {emptyt} (NoContext))))
    (check (= (Get tup2 2) (Const (Int 2) {emptyt} (NoContext))))
    (check (= (Get tup2 3) (Const (Int 3) {emptyt} (NoContext))))
    (check (= 4 (tuple-length tup2)))
    (fail (check (Get tup2 4)))
    "
    );
    egglog_test(
        &build,
        &check,
        vec![],
        Value::Tuple(vec![]),
        Value::Tuple(vec![]),
        vec![],
    )
}

#[test]
fn test_expr_size() -> crate::Result {
    use crate::ast::*;
    let pureloop = dowhile(
        single(int(1)),
        parallel!(
            less_than(getat(0), int(3)),
            get(switch!(int(2); parallel!(int(4), int(5))), 0)
        ),
    )
    .with_arg_types(emptyt(), tuplet!(intt()));
    let build: String = format!("(let loop {})", pureloop);

    let check = "(check (= 9 (Expr-size loop)))";
    egglog_test(
        build.as_str(),
        check,
        vec![],
        Value::Tuple(vec![]),
        Value::Tuple(vec![]),
        vec![],
    )
}

#[test]
fn test_sub_tuple() -> crate::Result {
    use crate::ast::*;
    let ty = tuplet!(intt(), intt(), intt());
    let arg = arg().with_arg_types(ty.clone(), ty.clone());
    let build = format!("(let expr (SubTuple {} 0 3))", arg);
    let out = concat(
        single(getat(0)),
        concat(single(getat(1)), single(getat(2))),
    )
    .with_arg_types(ty.clone(), ty);

    let check = format!("(check (= expr {}))", out);
    egglog_test(
        build.as_str(),
        &check,
        vec![],
        Value::Tuple(vec![]),
        Value::Tuple(vec![]),
        vec![],
    )
}
