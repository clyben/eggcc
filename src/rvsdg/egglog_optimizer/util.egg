;; ####################################
;; implementation of PassThroughArguments
;; Creates a vec of arguments
;; (vec-of (Arg 0) (Arg 1) ...) with length i
(function PassThroughArguments (i64) VecOperand :unextractable)



;; (how many arguments to generate, vector so far)
(function PassThroughArgumentsHelper (i64 VecOperand) VecOperand :unextractable)

(rewrite (PassThroughArguments i)
  (PassThroughArgumentsHelper i (VO (vec-of)))
  :ruleset subst)

(rule ((= lhs (PassThroughArgumentsHelper i (VO rest)))
       (< (vec-length rest) i))
      ((union lhs
        (PassThroughArgumentsHelper i
            (VO (vec-push rest (Arg (vec-length rest)))))))
      :ruleset subst)

(rule ((= lhs (PassThroughArgumentsHelper i (VO rest)))
       (= (vec-length rest) i))
      ((union lhs (VO rest)))
      :ruleset subst)



;; Project each argument out of a body
(function BodyToVecOperand (i64 Body) VecOperand :unextractable)
;; current index, body length, body, and vector so far
(function BodyToVecOperandHelper (i64 i64 Body VecOperandBase) VecOperand :unextractable)

(rewrite (BodyToVecOperand body-len body)
         (BodyToVecOperandHelper 0 body-len body (vec-of)) :ruleset subst)
(rule
  ((= helper (BodyToVecOperandHelper index body-len body so-far))
   (< index body-len))
  ((union helper
          (BodyToVecOperandHelper (+ index 1) body-len body
                                  (vec-push so-far 
                                   (Project index body)))))
  :ruleset subst)

(rule
  ((= helper (BodyToVecOperandHelper index body-len body so-far))
   (= index body-len))
  ((union helper (VO so-far)))
  :ruleset subst)

;; Converts (vec n1 n2 ...) to (vec (Arg n1) (Arg n2) ...)
(function MapArgOverVec (VecInt) VecOperand :unextractable)
(function MapArgOverVecHelper (VecIntBase i64 VecOperandBase) VecOperand :unextractable)
(let vec-operand-empty (vec-pop (vec-of (Arg 0))))
(rewrite (MapArgOverVec (VI vec)) 
         (MapArgOverVecHelper vec 0 vec-operand-empty)
      :ruleset subst)
(rewrite (MapArgOverVecHelper vec i curr)
         (MapArgOverVecHelper vec (+ i 1) (vec-push curr (Arg (vec-get vec i))))
      :when ((< i (vec-length vec)))
      :ruleset subst)
(rewrite (MapArgOverVecHelper vec i curr)
         (VO curr)
      :when ((= i (vec-length vec)))
      :ruleset subst)

(function VecAppend (VecOperand VecOperand) VecOperand)
(rewrite (VecAppend (VO v1) (VO v2)) 
         (VO (vec-append v1 v2))
      :ruleset subst)

;; Converts (vec n1 n2 ...) to (map-of n1 (Arg offset) n2 (Arg (+ offset 1)) ...)
;;                           vec    offset
(function BuildSubstMapFrom (VecInt i64) MapIntOperand :unextractable)
(function BuildSubstMapFromHelper (VecIntBase i64 i64 MapIntOperandBase) MapIntOperand :unextractable)
(rewrite (BuildSubstMapFrom (VI vi) offset) 
         ;; because type checking for actions is broken, we 
         ;; have to make sure there is only one Map sort right now
         (BuildSubstMapFromHelper vi offset 0 (map-empty))
         :ruleset subst)
(rewrite (BuildSubstMapFromHelper vi offset i m)
         (BuildSubstMapFromHelper vi offset (+ i 1) (map-insert m (vec-get vi i) (Arg (+ i offset))))
      :when ((< i (vec-length vi)))
      :ruleset subst)
(rewrite (BuildSubstMapFromHelper vi offset i m)
         (MIO m)
      :when ((= i (vec-length vi)))
      :ruleset subst)
