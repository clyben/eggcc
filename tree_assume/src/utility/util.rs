#[cfg(test)]
use crate::{egglog_test, interpreter::Value};

#[test]
fn test_list_util() -> crate::Result {
    let build = "
		(let list (Cons (Const (Int 0))
                  (Cons (Const (Int 1))
                  (Cons (Const (Int 2))
                  (Cons (Const (Int 3))
                  (Cons (Const (Int 4)) (Nil)))))))
		(let expr (Switch (Const (Int 1)) list))
	";
    let check = "
		(check (= (ListExpr-ith list 1) (Const (Int 1))))
        (check (= (ListExpr-ith list 4) (Const (Int 4))))
        (check (= (ListExpr-length list) 5))
	";
    egglog_test(
        &build,
        &check,
        vec![],
        Value::Tuple(vec![]),
        Value::Tuple(vec![]),
    )
}

#[test]
fn append_test() -> crate::Result {
    let build = "
        (let appended
            (Append
                (Cons (Const (Int 0)) (Cons (Const (Int 1)) (Nil)))
                (Const (Int 2))))
    ";

    let check = "
        (check (
            =
            (Cons (Const (Int 0)) (Cons (Const (Int 1)) (Cons (Const (Int 2)) (Nil))))
            appended
        ))
    ";
    egglog_test(
        &build,
        &check,
        vec![],
        Value::Tuple(vec![]),
        Value::Tuple(vec![]),
    )
}