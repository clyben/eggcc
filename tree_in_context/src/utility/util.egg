(function ListExpr-length (ListExpr) i64)
(function ListExpr-ith (ListExpr i64) Expr :unextractable)
(function ListExpr-suffix (ListExpr i64) ListExpr :unextractable)
(function Append (ListExpr Expr) ListExpr :unextractable)

(rule ((Switch pred branch)) ((union (ListExpr-suffix branch 0) branch)) :ruleset always-run)

(rule ((= (ListExpr-suffix top n) (Cons hd tl)))
    ((union (ListExpr-ith top n) hd)
     (union (ListExpr-suffix top (+ n 1)) tl)) :ruleset always-run)

(rule ((= (ListExpr-suffix list n) (Nil)))
    ((set (ListExpr-length list) n)) :ruleset always-run)

(rewrite (Append (Cons a b) e)
   (Cons a (Append b e))
   :ruleset always-run)
(rewrite (Append (Nil) e)
   (Cons e (Nil))
   :ruleset always-run)

(function tuple-ith (Expr i64) Expr :unextractable)
(function tuple-length (Expr) i64 :unextractable)

(rule ((HasType expr (TupleT tl))
       (= len (TypeList-length tl)))
      ((set (tuple-length expr) len)) :ruleset always-run)

(rule ((Single expr)) ((union (tuple-ith (Single expr) 0) expr)) :ruleset always-run)
(rule ((Concat ord expr1 expr2)
       (= ith (tuple-ith expr1 i)))
       ((union (tuple-ith (Concat ord expr1 expr2) i) ith)) :ruleset always-run)

(rule ((Concat ord expr1 expr2)
       (= len1 (tuple-length expr1))
       (= ith (tuple-ith expr2 i)))
       ((union (tuple-ith (Concat ord expr1 expr2) (+ len1 i)) ith)) :ruleset always-run)

;; This Expr size actually do the same thing as cost function, can be removed afterwards.
(function Expr-size (Expr) i64 :merge (min old new))
(function ListExpr-size (ListExpr) i64 :merge (min old new))

(rule ((= expr (Function name tyin tyout out)) 
       (= sum (Expr-size out))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (Const n ty))) 
      ((set (Expr-size expr) 1))  :ruleset always-run)

(rule ((= expr (Bop op x y)) 
       (= sum (+ (Expr-size y) (Expr-size x)))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (Uop op x)) 
       (= sum (Expr-size x))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (Get tup i)) 
       (= sum (Expr-size tup))) 
      ((set (Expr-size expr) sum)) :ruleset always-run)

(rule ((= expr (Concat order x y)) 
       (= sum (+ (Expr-size y) (Expr-size x)))) 
      ((set (Expr-size expr) sum)) :ruleset always-run)

(rule ((= expr (Single x)) 
       (= sum (Expr-size x))) 
      ((set (Expr-size expr) sum)) :ruleset always-run)

(rule ((= expr (Switch pred branches)) 
       (= sum (+ (ListExpr-size branches) (Expr-size pred)))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (If pred then else)) 
       (= sum (+ (Expr-size else) (+ (Expr-size then) (Expr-size pred))))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (DoWhile in pred-and-output)) 
       (= sum (+ (Expr-size pred-and-output) (Expr-size in)))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (Let in out)) 
       (= sum (+ (Expr-size out) (Expr-size in)))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (Arg ty))) 
      ((set (Expr-size expr) 1)) :ruleset always-run)

(rule ((= expr (Call func arg)) 
       (= sum (Expr-size arg))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((Empty ty)) ((set (Expr-size (Empty ty)) 0))  :ruleset always-run)

(rule ((= expr (Cons hd tl)) 
       (= sum (+ (ListExpr-size tl) (Expr-size hd)))) 
      ((set (ListExpr-size expr) sum)) :ruleset always-run)

(rule ((Nil)) 
      ((set (ListExpr-size (Nil)) 0))  :ruleset always-run)

(rule ((= expr (Alloc e ty)) 
        (= sum (Expr-size e))) 
        ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (InContext assumption e)) 
       (= sum (Expr-size e))) 
      ((set (Expr-size expr) sum)) :ruleset always-run)


(function tuple-of-get-arg (Type i64) Expr :unextractable)
(function tuple-of-get-arg-helper (Type i64 i64) Expr :unextractable)

(rewrite (tuple-of-get-arg ty n) (tuple-of-get-arg-helper ty n 0) :ruleset always-run)
(rewrite (tuple-of-get-arg-helper ty 1 x) (Single (Get (Arg ty) x)) :ruleset always-run)
(rule ((= f (tuple-of-get-arg-helper ty a b)) (> a 1))
      ((union f 
         (Concat par
               (Get (Arg ty) b) 
               (tuple-of-get-arg-helper ty (- a 1) (+ b 1))))) :ruleset always-run)