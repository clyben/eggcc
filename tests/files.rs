use eggcc::util::Run;
use insta::assert_snapshot;
use libtest_mimic::Trial;

fn generate_tests(glob: &str) -> Vec<Trial> {
    let mut trials = vec![];
    let mut mk_trial = |run: Run| {
        trials.push(Trial::test(run.name(), move || {
            let result = run.run();

            if let Some(interpreted) = result.result_interpreted {
                assert_eq!(result.original_interpreted, interpreted);
            } else {
                assert_snapshot!(run.name(), result.visualization);
            }
            Ok(())
        }))
    };

    for entry in glob::glob(glob).unwrap() {
        let f = entry.unwrap();
        let name = f
            .file_stem()
            .unwrap()
            .to_string_lossy()
            .replace(['.', '-', ' '], "_");

        // TODO optimizer doesn't support these yet
        let banned = ["queens_func", "unstructured", "implicit_return", "fib"];
        if banned.iter().any(|b| name.contains(b)) || f.to_str().unwrap().contains("failing") {
            continue;
        }

        for run in Run::all_configurations_for(f) {
            mk_trial(run);
        }
    }

    trials
}

fn main() {
    let args = libtest_mimic::Arguments::from_args();
    let tests = generate_tests("tests/**/*.bril");
    libtest_mimic::run(&args, tests).exit();
}
