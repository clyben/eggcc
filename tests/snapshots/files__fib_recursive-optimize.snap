---
source: tests/files.rs
expression: visualization.result
---
@fac(v0: int): int {
.v1_:
  v2_: int = const 0;
  v3_: bool = eq v0 v2_;
  v4_: int = const 1;
  v5_: int = id v4_;
  br v3_ .v6_ .v7_;
.v7_:
  v8_: bool = eq v0 v4_;
  br v8_ .v9_ .v10_;
.v9_:
  v11_: bool = eq v2_ v2_;
  v12_: int = id v0;
  br v11_ .v13_ .v14_;
.v14_:
  v15_: bool = eq v0 v2_;
  br v15_ .v16_ .v17_;
.v16_:
  v18_: bool = eq v2_ v2_;
  v19_: int = id v2_;
  br v18_ .v20_ .v21_;
.v21_:
  v22_: bool = eq v2_ v2_;
  br v22_ .v23_ .v24_;
.v23_:
  v25_: int = call @fac v2_;
  v26_: int = id v2_;
.v27_:
  v19_: int = id v26_;
.v20_:
  v28_: int = id v2_;
.v29_:
  v12_: int = id v28_;
.v13_:
  v30_: int = id v0;
.v31_:
  v5_: int = id v30_;
  ret v5_;
.v24_:
  v32_: int = const -1;
  v33_: int = sub v32_ v2_;
  v34_: int = call @fac v32_;
  v35_: int = call @fac v33_;
  v36_: int = add v34_ v35_;
  v26_: int = id v36_;
  jmp .v27_;
.v17_:
  v37_: int = const -2;
  v38_: bool = eq v2_ v37_;
  v39_: int = const -1;
  v40_: bool = eq v2_ v39_;
  v41_: int = id v0;
  br v40_ .v42_ .v43_;
.v43_:
  v44_: bool = eq v0 v39_;
  br v44_ .v45_ .v46_;
.v45_:
  v47_: int = call @fac v2_;
  v48_: int = id v39_;
.v49_:
  v41_: int = id v48_;
.v42_:
  v50_: int = id v0;
  br v38_ .v51_ .v52_;
.v52_:
  v53_: bool = eq v0 v37_;
  br v53_ .v54_ .v55_;
.v54_:
  v56_: int = call @fac v2_;
  v57_: int = id v0;
.v58_:
  v50_: int = id v57_;
.v51_:
  v59_: int = add v41_ v50_;
  v28_: int = id v59_;
  jmp .v29_;
.v55_:
  v60_: int = const -4;
  v61_: int = const -3;
  v62_: int = call @fac v61_;
  v63_: int = call @fac v60_;
  v64_: int = add v62_ v63_;
  v57_: int = id v64_;
  jmp .v58_;
.v46_:
  v65_: int = const -3;
  v66_: int = const -2;
  v67_: int = call @fac v66_;
  v68_: int = call @fac v65_;
  v69_: int = add v67_ v68_;
  v48_: int = id v69_;
  jmp .v49_;
.v10_:
  v70_: int = sub v0 v4_;
  v71_: int = sub v70_ v4_;
  v72_: bool = eq v2_ v71_;
  v73_: bool = eq v2_ v70_;
  v74_: int = id v4_;
  br v73_ .v75_ .v76_;
.v76_:
  v77_: bool = eq v4_ v70_;
  br v77_ .v78_ .v79_;
.v78_:
  v80_: bool = eq v2_ v2_;
  v81_: int = id v70_;
  br v80_ .v82_ .v83_;
.v83_:
  v84_: bool = eq v2_ v70_;
  br v84_ .v85_ .v86_;
.v85_:
  v87_: int = call @fac v2_;
  v88_: int = id v2_;
.v89_:
  v81_: int = id v88_;
.v82_:
  v90_: int = id v70_;
.v91_:
  v74_: int = id v90_;
.v75_:
  v92_: int = id v4_;
  br v72_ .v93_ .v94_;
.v94_:
  v95_: bool = eq v4_ v71_;
  br v95_ .v96_ .v97_;
.v96_:
  v98_: bool = eq v2_ v2_;
  v99_: int = id v71_;
  br v98_ .v100_ .v101_;
.v101_:
  v102_: bool = eq v2_ v71_;
  br v102_ .v103_ .v104_;
.v103_:
  v105_: int = call @fac v2_;
  v106_: int = id v2_;
.v107_:
  v99_: int = id v106_;
.v100_:
  v108_: int = id v71_;
.v109_:
  v92_: int = id v108_;
.v93_:
  v110_: int = add v74_ v92_;
  v30_: int = id v110_;
  jmp .v31_;
.v104_:
  v111_: int = const -2;
  v112_: int = const -1;
  v113_: int = call @fac v112_;
  v114_: int = call @fac v111_;
  v115_: int = add v113_ v114_;
  v106_: int = id v115_;
  jmp .v107_;
.v97_:
  v116_: int = sub v71_ v4_;
  v117_: int = sub v116_ v4_;
  v118_: bool = eq v117_ v2_;
  v119_: bool = eq v116_ v2_;
  v120_: int = id v4_;
  br v119_ .v121_ .v122_;
.v122_:
  v123_: bool = eq v116_ v4_;
  br v123_ .v124_ .v125_;
.v124_:
  v126_: int = call @fac v2_;
  v127_: int = id v4_;
.v128_:
  v120_: int = id v127_;
.v121_:
  v129_: int = id v4_;
  br v118_ .v130_ .v131_;
.v131_:
  v132_: bool = eq v117_ v4_;
  br v132_ .v133_ .v134_;
.v133_:
  v135_: int = call @fac v2_;
  v136_: int = id v4_;
.v137_:
  v129_: int = id v136_;
.v130_:
  v138_: int = add v120_ v129_;
  v108_: int = id v138_;
  jmp .v109_;
.v134_:
  v139_: int = sub v117_ v4_;
  v140_: int = sub v139_ v4_;
  v141_: int = call @fac v139_;
  v142_: int = call @fac v140_;
  v143_: int = add v141_ v142_;
  v136_: int = id v143_;
  jmp .v137_;
.v125_:
  v144_: int = sub v116_ v4_;
  v145_: int = sub v144_ v4_;
  v146_: int = call @fac v144_;
  v147_: int = call @fac v145_;
  v148_: int = add v146_ v147_;
  v127_: int = id v148_;
  jmp .v128_;
.v86_:
  v149_: int = const -2;
  v150_: int = const -1;
  v151_: int = call @fac v150_;
  v152_: int = call @fac v149_;
  v153_: int = add v151_ v152_;
  v88_: int = id v153_;
  jmp .v89_;
.v79_:
  v154_: int = sub v70_ v4_;
  v155_: int = sub v154_ v4_;
  v156_: bool = eq v155_ v2_;
  v157_: bool = eq v154_ v2_;
  v158_: int = id v4_;
  br v157_ .v159_ .v160_;
.v160_:
  v161_: bool = eq v154_ v4_;
  br v161_ .v162_ .v163_;
.v162_:
  v164_: int = call @fac v2_;
  v165_: int = id v154_;
.v166_:
  v158_: int = id v165_;
.v159_:
  v167_: int = id v4_;
  br v156_ .v168_ .v169_;
.v169_:
  v170_: bool = eq v155_ v4_;
  br v170_ .v171_ .v172_;
.v171_:
  v173_: int = call @fac v2_;
  v174_: int = id v4_;
.v175_:
  v167_: int = id v174_;
.v168_:
  v176_: int = add v158_ v167_;
  v90_: int = id v176_;
  jmp .v91_;
.v172_:
  v177_: int = sub v155_ v4_;
  v178_: int = sub v177_ v4_;
  v179_: int = call @fac v177_;
  v180_: int = call @fac v178_;
  v181_: int = add v179_ v180_;
  v174_: int = id v181_;
  jmp .v175_;
.v163_:
  v182_: int = sub v154_ v4_;
  v183_: int = sub v182_ v4_;
  v184_: int = call @fac v182_;
  v185_: int = call @fac v183_;
  v186_: int = add v184_ v185_;
  v165_: int = id v186_;
  jmp .v166_;
.v6_:
  ret v5_;
}
@main {
.v0_:
  v1_: int = const 2;
  v2_: int = const 0;
  v3_: bool = eq v1_ v2_;
  v4_: int = const 1;
  v5_: int = id v4_;
  br v3_ .v6_ .v7_;
.v7_:
  v8_: bool = eq v1_ v4_;
  br v8_ .v9_ .v10_;
.v9_:
  v11_: bool = eq v2_ v2_;
  v12_: int = id v1_;
  br v11_ .v13_ .v14_;
.v14_:
  v15_: bool = eq v1_ v2_;
  br v15_ .v16_ .v17_;
.v16_:
  v18_: int = call @fac v2_;
  v19_: int = id v1_;
.v20_:
  v12_: int = id v19_;
.v13_:
  v21_: int = id v1_;
.v22_:
  v5_: int = id v21_;
  print v5_;
  ret;
.v17_:
  v23_: int = const -1;
  v24_: int = sub v23_ v1_;
  v25_: int = call @fac v23_;
  v26_: int = call @fac v24_;
  v27_: int = add v25_ v26_;
  v19_: int = id v27_;
  jmp .v20_;
.v10_:
  v28_: bool = eq v2_ v2_;
  v29_: bool = eq v2_ v4_;
  v30_: int = id v4_;
  br v29_ .v31_ .v32_;
.v32_:
  v33_: bool = eq v4_ v4_;
  br v33_ .v34_ .v35_;
.v34_:
  v36_: int = call @fac v2_;
  v37_: int = id v4_;
.v38_:
  v30_: int = id v37_;
.v31_:
  v39_: int = id v4_;
  br v28_ .v40_ .v41_;
.v41_:
  v42_: bool = eq v2_ v4_;
  br v42_ .v43_ .v44_;
.v43_:
  v45_: int = call @fac v2_;
  v46_: int = id v2_;
.v47_:
  v39_: int = id v46_;
.v40_:
  v48_: int = add v30_ v39_;
  v21_: int = id v48_;
  jmp .v22_;
.v44_:
  v49_: int = const -2;
  v50_: int = const -1;
  v51_: int = call @fac v50_;
  v52_: int = call @fac v49_;
  v53_: int = add v51_ v52_;
  v46_: int = id v53_;
  jmp .v47_;
.v35_:
  v54_: int = const -1;
  v55_: int = call @fac v2_;
  v56_: int = call @fac v54_;
  v57_: int = add v55_ v56_;
  v37_: int = id v57_;
  jmp .v38_;
.v6_:
  print v5_;
}
